Fixing lost state changes with ui.timer

rodja opened the pull request.

falkoschindler commented:
Looks good. But I wonder if we should treat `_run_in_loop` and `_run_once` more consistently. The user might want to implement a short delay which could be invoked before handshake.
And I wonder if we could move the handshake out of the while loop and do it only once before. 
Either way, there is the potential danger of a handshake never happening because the client disconnected, leaving an endless loop checking for `environ` ten times a second. But this is probably a general problem with our `handshake` we should discuss separately.

rodja commented:
Good suggestions! I moved the handshake-awaiting out of the loop and implemented the same behavior vor `_run_once`. I also parameterized the test to test both cases.
